### PR TITLE
5814 sass lint

### PR DESCRIPTION
**Pull request by votemike**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-5814

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Ds2013/Core/Gel/Base/_typography-generic.scss [hunk 1]
```diff
@@ -143,14 +147,16 @@
     }
 }
 
-@include gel-text-sizing('', map-get($gel-typography, 'basic')...);
+$map-get: map-get($gel-typography, 'basic');
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Base/_typography-generic.scss]: As this variable is going to end up in the global namespace, can it be named a bit better? `$basic-typography`?

I know Gonzales can't parse the old code, I'm unsure if we want to work around it, or stick this file in the list of "ignored because wrong invalids" and raise the issue upstream

File: src/Ds2013/Core/Gel/Base/_typography-generic.scss [hunk 2]
```diff
@@ -143,14 +147,16 @@
...
+@include gel-text-sizing('', $map-get...);
 
 /**
  * Generate media-query scoped font sizings for each breakpoint name defined
  */
 @each $item in $gel-typography-breakpoints {
     @include mq-range($item) {
-        @include gel-text-sizing('#{$item}-', map-get($gel-typography, $item)...);
+        $map-get: map-get($gel-typography, $item);
+        @include gel-text-sizing('#{$item}-', $map-get...) ;
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Base/_typography-generic.scss]: Same query about "Gonzales chokes on this, should we ignore the whole file"?

If we fix can we name the variable a bit better?

File: src/Ds2013/Core/Gel/Generic/_mixins.scss [hunk 3]
```diff
@@ -6,23 +6,24 @@
 
 @mixin mq-range($name, $aliases: $mq-aliases) {
     @if map-has-key($aliases, $name) {
-       @include mq(map-get($aliases, $name)...) {
+        $map-get: map-get($aliases, $name);
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Generic/_mixins.scss]: Better name for the variable?

File: src/Ds2013/Core/Gel/Generic/_mixins.scss [hunk 4]
```diff
@@ -6,23 +6,24 @@
...
-        @warn "Range '#{$name}' was not found in the list of defined ranges";
+        @warn 'Range \'#{$name}\' was not found in the list of defined ranges';
     }
 }
 
 // These mixins only exist to help the migration process. Give a helpful error when you attempt to use breakup-block
 // or breakup-breakpoint with information on how to fix it
 @mixin breakup-block($name) {
-    @error "You used breakup-block. Blocks with the name of 'basic' no longer need to be included within a mixin, their content can go directly the scss file";
+    @error 'You used breakup-block. Blocks with the name of \'basic\' no longer need to be included within a mixin, their content can go directly the scss fil';
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Generic/_mixins.scss]: Accidentally deleted the e at the end of the line

File: src/Ds2013/Core/Gel/Generic/_variables.scss [hunk 5]
```diff
@@ -40,7 +40,7 @@ $gel-class-prefix: unquote(if($gel-use-silent-classes, "%", "."));
  * bpb2   | 480  | 599  | Landscape smartphones
  * bpw    | 600  | -    | All wide displays
  * bpw1   | 600  | 1007 | Synonym for bp3 for Tablets
- * bpw2   | 770  | 1007 | Landscape 10" tablets
+ * bpw2   | 770  | 1007 | Landscape 10' tablets
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Generic/_variables.scss]: 10 foot tablets!

File: src/Ds2013/Core/Gel/Objects/_gelicon.scss [hunk 6]
```diff
@@ -76,14 +89,16 @@ svg.gelicon--alpha, svg.gelicon--beta, svg.gelicon--gamma, svg.gelicon--delta, s
     }
 }
 
-@include svg-sizing('', map-get($gel-typography, 'basic')...);
+$map-get: map-get($gel-typography, 'basic');
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Objects/_gelicon.scss]: Better variable name?

Or a candidate for ignoring the whole file?

File: src/Ds2013/Core/Gel/Objects/_gelicon.scss [hunk 7]
```diff
@@ -76,14 +89,16 @@ svg.gelicon--alpha, svg.gelicon--beta, svg.gelicon--gamma, svg.gelicon--delta, s
...
+@include svg-sizing('', $map-get...);
 
 /**
  * Generate media-query scoped gel-icon sizings that correspond to our font sizings for each breakpoint name defined
  */
 @each $item in $gel-typography-breakpoints {
     @include mq-range($item) {
-        @include svg-sizing('#{$item}-', map-get($gel-typography, $item)...);
+        $map-get: map-get($gel-typography, $item);
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Objects/_gelicon.scss]: Better variable name?

File: src/Ds2013/Core/Gel/Objects/_widths.scss [hunk 8]
```diff
@@ -16,64 +16,151 @@
...
 
-
 /**
  * Grid width sizing ratios, these have grid-agnostic names as they can also
  * be used for items within a grid, such as on .media__img
  */
 @mixin gel-widths-setup($suffix: '') {
     // Whole
-    .\31\/1#{$suffix} { width: (1/1) * 100%; } // .1/1
+    .\31\/1#{$suffix} {
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/Objects/_widths.scss]: Rather than rewrite this whole block would it be better to temporarily ignore the one-declaration-per-line?

File: src/Ds2013/Core/Gel/_gel.scss [hunk 9]
```diff
@@ -4,12 +4,9 @@
  * GEL styling to be included as standard in the common stylesheet that always
  * gets loaded.
  */
-
```
Comment (BPScott) [inline src/Ds2013/Core/Gel/_gel.scss]: Is this empty spacing causing a problem? it seemed to be acceptable in the files in `app`

File: src/Ds2013/Core/Helpers/_widths.scss [hunk 10]
```diff
@@ -12,16 +12,20 @@
...
+    .\31\/6\\@bpw {
+        width: 976px * 0.16666;
+    }
 }
 
 // For letting you undo grid-unbounded at a given breakpoint
 @each $name in $gel-widths-breakpoints {
     @include mq-range($name) {
-           .grid--unbounded\\@#{$name} {
+        .grid--unbounded\\@ #{$name} {
```
Comment (BPScott) [inline src/Ds2013/Core/Helpers/_widths.scss]: Don't add a space after the `@`

File: src/DsAmen/Atom/_typography.scss [hunk 11]
```diff
@@ -71,7 +71,8 @@ $type-settings: (
 }
 
 @each $class-name, $groups in $type-settings {
-    @include _typography-group($class-name, map-get($groups, 'group-a')...);
+    $map-get: map-get($groups, 'group-a');
```
Comment (BPScott) [inline src/DsAmen/Atom/_typography.scss]: Better variable name?

Or ignore the file?

File: src/DsAmen/Atom/_typography.scss [hunk 12]
```diff
@@ -81,7 +82,8 @@ $type-settings: (
 @each $gel-group, $amen-mq in ('group-b': 'gel1b', 'group-d': 'gel3') {
     @include mq($from: $amen-mq) {
         @each $class-name, $groups in $type-settings {
-            @include _typography-group($class-name, map-get($groups, $gel-group)...)
+            $map-get: map-get($groups, $gel-group);
```
Comment (BPScott) [inline src/DsAmen/Atom/_typography.scss]: better variable name or ignore the file

File: src/DsAmen/Core/Vendor/_mq.scss [hunk 13]
```diff
@@ -1,5 +1,5 @@
-@charset "UTF-8"; // Fixes an issue where Ruby locale is not set properly
```
Comment (BPScott) [inline src/DsAmen/Core/Vendor/_mq.scss]: Vendor folder should be ignored, don't make changes in here.

File: src/DsAmen/Core/Vendor/_normalize.scss [hunk 14]
```diff
@@ -7,17 +7,17 @@
  */
 
 html {
-  font-family: sans-serif; /* 1 */
```
Comment (BPScott) [inline src/DsAmen/Core/Vendor/_normalize.scss]: Revert the changes as Vendor should be ignored

File: src/DsAmen/Core/_fractions.scss [hunk 15]
```diff
@@ -8,72 +8,246 @@
     // \31 = 1, \32 = 2 etc.
 
     // whole
-    .\31\/1#{$suffix} { width: (1/1) * 100%; } // .1/1
```
Comment (BPScott) [inline src/DsAmen/Core/_fractions.scss]: Rather than rewrite this whole block would it be better to temporarily ignore the one-declaration-per-line rule?

File: src/Ds2013/Core/Gel/Objects/_widths.scss (hunk 8 above)
Comment (votemike) [inline src/Ds2013/Core/Gel/Objects/_widths.scss]: Can do. Up to you.

File: src/DsAmen/Core/_fractions.scss (hunk 15 above)
Comment (votemike) [inline src/DsAmen/Core/_fractions.scss]: Up to you.

File: src/Ds2013/Core/Gel/Objects/_widths.scss (hunk 8 above)
Comment (BPScott) [inline src/Ds2013/Core/Gel/Objects/_widths.scss]: I think the current format is more readable (and has the added bonus of reducing diff noise)

Comment (votemike) [inline src/Ds2013/Core/Gel/Objects/_widths.scss]: Done

File: script/ci-docker [hunk 16]
```diff
@@ -38,6 +38,9 @@ vendor/bin/phpunit || RESULT=1
 # Run PHPCS
 vendor/bin/phpcs -n --report=checkstyle --report-file=build/reports/checkstyle.xml || RESULT=1
 
+# Run sass-lint
+node_modules/.bin/sass-lint --no-exit -v || RESULT=1
```
Comment (tatitati) [inline script/ci-docker]: Is ok I guess, but I think we should try to keep all the frontend build stuff in yarn gulp instead of running in different places.

In there we have all the stuff required to build/check/compile the frontend. Even one day we can add "jslint" as well.

As part of the frontend build we might use gulp-sass-lint, which accepts the same options we have in sass-lint.yml

Comment (votemike) [inline script/ci-docker]: What sort of thing are you thinking? Some sort of `gulp test`? Should SASS also be checked before compiling into CSS and bad SASS also make any gulp builds fail?

Comment (tatitati) [inline script/ci-docker]: gulp lint would be nice. And yes, I think in local it should fail if the css is wrong (as we make with phpcs or can happen with jslint one day). But this can be flexible. In script/test we can have
```
vendor/bin/phpunit
vendor/bin/phpcs
gulp tests (in the future)
gulp lint 
```
I wouldn't compile sass all the time, script/test should be for testing the code.

File: package.json [hunk 17]
```diff
@@ -15,11 +15,13 @@
         "gulp-sass-error": "^1.0.5",
         "gulp-sourcemaps": "^2.4.1",
         "lazysizes": "^3.0.0",
+        "sass-lint": "^1.10.2",
         "sass-mq": "^3.3.2"
     },
     "scripts": {
         "gulp": "gulp default",
         "gulp-ci": "gulp distribution",
+        "lint": "node_modules/.bin/sass-lint --no-exit -v",
```
Comment (votemike) [inline package.json]: Do we want this to be `"lint"` or `"lint:sass"`?

Comment (votemike) [inline package.json]: Do we want this to be `"node_modules/.bin/sass-lint --no-exit -v"` or `"gulp lint"`?

Comment (tatitati) [inline package.json]: I would say lint:sass

File: package.json [hunk 18]
```diff
@@ -21,6 +21,7 @@
...
```
Comment (tatitati) [inline package.json]: I would put in gulpfile.js something like this:
```javascript
gulp.task('lint', ['sass:lint', 'js:lint']);
gulp.task('test', ['js:test']);
```

The default task sass then in gulpfile also would look like this:
```javascript
gulp.task('sass', ['sass:lint', 'sass:clean', 'sass:lint'], function() {
....
```

And in our package.json this:
```json
"scripts": {
    "gulp": "gulp default",
    "gulp-ci": "gulp distribution",

    "gulp-lint": "gulp lint",
    "gulp-test": "gulp test",
}
```

And in our script/test this:
```bash
vendor/bin/phpunit
vendor/bin/phpcs
yarn gulp-lint
yarn gulp-test
```
